pkg/runtime/dispatcher.go

```diff
@@ -0,0 +1,443 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	rpc "github.com/googleapis/googleapis/google/rpc"
+	"github.com/hashicorp/go-multierror"
+	"github.com/opentracing/opentracing-go"
+	tracelog "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"istio.io/mixer/pkg/adapter"
+	adptTmpl "istio.io/mixer/pkg/adapter/template"
+	"istio.io/mixer/pkg/aspect"
+	"istio.io/mixer/pkg/attribute"
+	cpb "istio.io/mixer/pkg/config/proto"
+	"istio.io/mixer/pkg/expr"
+	"istio.io/mixer/pkg/pool"
+	"istio.io/mixer/pkg/status"
+	"istio.io/mixer/pkg/template"
+)
+
+// Action is the runtime representation of configured action - cpb.Action.
+// Configuration is processed to hydrate instance names to Instances and handler.
+type Action struct {
+	// ready to use handler
+	handler adapter.Handler
+	// configuration for the handler
+	handlerConfig *cpb.Handler
+	// instanceConfigs to dispatch to the handler.
+	// instanceConfigs must belong to the same template.
+	instanceConfig []*cpb.Instance
+}
+
+// Resolver represents the current snapshot of the configuration database
+// and associated, initialized handlers.
+type Resolver interface {
+	// Resolve resolves configuration to a list of actions.
+	Resolve(bag attribute.Bag, variety adptTmpl.TemplateVariety) ([]*Action, error)
+}
+
+// Newdispatcher creates a new dispatcher.
+func Newdispatcher(mapper expr.Evaluator, templateRepo template.Repository,
+	rt Resolver, gp *pool.GoroutinePool) Dispatcher {
+	m := &dispatcher{
+		mapper:       mapper,
+		templateRepo: templateRepo,
+		gp:           gp,
+	}
+	m.SetResolver(rt)
+	return m
+}
+
+// dispatcher is responsible for dispatching incoming API calls
+// to the configured adapters. It implements the Dispatcher interface.
+type dispatcher struct {
+	// mapper is the selector and expression evaluator.
+	// It is not directly used by dispatcher.
+	mapper expr.Evaluator
+
+	// Repository of templates that are compiled in this Mixer.
+	// It provides a way to perform template operations.
+	templateRepo template.Repository
+
+	// <resolver>
+	resolver atomic.Value
+
+	// gp is used to dispatch multiple adapters concurrently.
+	gp *pool.GoroutinePool
+}
+
+// SetState installs a new runtime state.
+func (m *dispatcher) SetResolver(rt Resolver) {
+	m.resolver.Store(rt)
+}
+
+// State gets the current runtime state.
+func (m *dispatcher) Resolver() Resolver {
+	return m.resolver.Load().(Resolver)
+}
+
+// Dispatcher dispatches incoming API calls to configured adapters.
```

Comment:
Can you move this to the top of the file, it's more customary to have interfaces up front.

Please update aspect-related comments to more modern phrasing.